src/python_testing/TC_FAN_3_5.py

```diff
@@ -21,207 +21,1090 @@
+    async def subscribe_to_attributes(self, handle_current_values: bool) -> None:
+        cluster = Clusters.FanControl
+        attr = cluster.Attributes
+
+        if hasattr(self, 'subscriptions') and self.subscriptions is not None:
+            for sub in self.subscriptions:
+                sub.cancel()
+
+        if not handle_current_values:
+            self.subscriptions = [
+                ClusterAttributeChangeAccumulator(cluster, attr.PercentSetting),
+                ClusterAttributeChangeAccumulator(cluster, attr.FanMode),
+                ClusterAttributeChangeAccumulator(cluster, attr.SpeedSetting)
+            ]
+        else:
+            self.subscriptions = [
+                ClusterAttributeChangeAccumulator(cluster, attr.PercentSetting),
+                ClusterAttributeChangeAccumulator(cluster, attr.PercentCurrent),
+                ClusterAttributeChangeAccumulator(cluster, attr.SpeedCurrent)
+            ]
+
+        for sub in self.subscriptions:
+            await sub.start(self.default_controller, self.dut_node_id, self.endpoint)
+
+    async def get_fan_modes(self, remove_auto: bool = False) -> None:
+        """Determine and store supported FanMode values from FanModeSequence attribute.
+
+        Args:
+            remove_auto (bool): If True, exclude the 'Auto' FanMode from the result.
+        """
+        # Read FanModeSequence attribute value
+        fan_mode_sequence_attr = Clusters.FanControl.Attributes.FanModeSequence
+        fm_enum = Clusters.FanControl.Enums.FanModeEnum
+        fms_enum = Clusters.FanControl.Enums.FanModeSequenceEnum
+        self.fan_mode_sequence = await self.read_setting(fan_mode_sequence_attr)
+
+        # Verify response contains a FanModeSequenceEnum
+        asserts.assert_is_instance(self.fan_mode_sequence, fms_enum,
+                                   f"[FC] FanModeSequence result isn't of enum type {fms_enum.__name__}")
+
+        fan_modes = None
+        if self.fan_mode_sequence == 0:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kMedium, fm_enum.kHigh]
+        elif self.fan_mode_sequence == 1:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kHigh]
+        elif self.fan_mode_sequence == 2:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kMedium, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 3:
+            fan_modes = [fm_enum.kOff, fm_enum.kLow, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 4:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh, fm_enum.kAuto]
+        elif self.fan_mode_sequence == 5:
+            fan_modes = [fm_enum.kOff, fm_enum.kHigh]
+
+        self.fan_modes = [f for f in fan_modes if not (remove_auto and f == fm_enum.kAuto)]
+
+    async def initialize_percent_setting_and_verify_affected_attribtutes(self, step: Clusters.FanControl.Commands.Step, handle_current_values: bool) -> int:
```

Comment:
It seems a bit odd to do this from the step command - why not just send in the percent setting you want and verify speed and mode off of that? It's either 0 or 100 for this test, right?